zoneinfo/src/types.rs

```diff
@@ -0,0 +1,650 @@
+                    day: DayOfMonth::Day(1),
+                },
+                time: QualifiedTime::Local(Time::default()),
+            });
+        };
+
+        let time = next_split(&mut splits, ctx)
+            .ok()
+            .map(|t| t.context_parse::<QualifiedTime>(ctx))
+            .transpose()?
+            .unwrap_or(QualifiedTime::Local(Time::default()));
+
+        ctx.exit();
+        Ok(Self { date, time })
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct Date {
+    pub year: i32,
+    pub month: Month,
+    pub day: DayOfMonth,
+}
+
+impl Date {
+    pub fn as_secs(&self) -> i64 {
+        let epoch_days = epoch_days_for_rule_date(self.year, self.month, self.day);
+        utils::epoch_seconds_for_epoch_days(epoch_days)
+    }
+}
+
+#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
+pub struct Time {
+    pub sign: Sign,
+    pub hour: i8,
+    pub minute: i8,
+    pub second: i8,
+}
+
+#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
+#[repr(i8)]
+pub enum Sign {
+    #[default]
+    Positive = 1,
+    Negative = -1,
+}
+
+impl Time {
+    pub const fn as_secs(&self) -> i64 {
+        (self.hour as i64 * 3600 + self.minute as i64 * 60 + self.second as i64) * self.sign as i64
+    }
+}
+
+impl TryFromStr<LineParseContext> for Time {
+    type Error = ZoneInfoParseError;
+    fn try_from_str(s: &str, ctx: &mut LineParseContext) -> Result<Self, Self::Error> {
+        ctx.enter("Time");
+        let (s, sign) = if let Some(stripped) = s.strip_prefix('-') {
+            (stripped, Sign::Negative)
+        } else {
+            (s, Sign::Positive)
+        };
+        if !s.contains(':') {
+            let hour = s.context_parse::<i8>(ctx)?;
+            ctx.exit();
+            return Ok(Time {
+                sign,
+                hour,
+                minute: 0,
+                second: 0,
+            });
+        }
+        let (hour, sub_hour) = s.split_once(':').ok_or(ZoneInfoParseError::UnknownValue(
+            ctx.line_number,
+            s.to_owned(),
+        ))?;
+        let hour = hour.context_parse::<i8>(ctx)?;
+        if !sub_hour.contains(':') {
+            let minute = sub_hour.context_parse::<i8>(ctx)?;
+            ctx.exit();
+            return Ok(Self {
+                sign,
+                hour,
+                minute,
+                second: 0,
+            });
+        }
+        let (minute, second) = sub_hour
+            .split_once(':')
+            .ok_or(ZoneInfoParseError::UnknownValue(
+                ctx.line_number,
+                s.to_owned(),
+            ))?;
+        let minute = minute.context_parse::<i8>(ctx)?;
+        let second = second.context_parse::<i8>(ctx)?;
+        ctx.exit();
+        Ok(Self {
+            sign,
+            hour,
+            minute,
+            second,
+        })
+    }
+}
+
+// ==== Rule types ====
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum ToYear {
+    Max,
+    Year(u16),
+}
```

Comment:
question: `Option<ToYear>` appears several places. Is there a distinction in behavior between `None` and `Some(ToYear::Max)` in those cases?

Reply:
Yep! `max` is a specified value, where `None` is an empty value / "-", which means the `ToYear` is the same as the FROM value.